module-system/sov-modules-api/src/lib.rs

```diff
@@ -333,10 +334,37 @@ pub trait ModuleInfo {
     fn dependencies(&self) -> Vec<&<Self::Context as Spec>::Address>;
 }
 
+/// A type that is addressable.
+pub trait Addressable {
+    /// Execution context.
+    type Context: Context;
+
+    /// Returns address of the instance.
+    fn address(&self) -> &<Self::Context as Spec>::Address;
+
+    /// Returns the addresses of all dependencies of this instance.
+    fn dependencies(&self) -> Vec<&<Self::Context as Spec>::Address>;
+}
+
+impl<T, C: Context> Addressable for T
```

Comment:
@vlopes11 can you please remind me why do we need `Addressable `. `ModuleInfo` has only one method more which shouldn't be problematic ` fn prefix(&self) -> Prefix;`? `Addressable` requires `ModuleInfo`

Reply:
Ahh this is because `ModuleInfo` will have a GasConfig? But the why GasConfig  goes to `ModuleInfo`?

Reply:
It's mainly because we sometimes take a module as a dynamic dispatch of `ModuleInfo`

https://github.com/Sovereign-Labs/sovereign-sdk/blob/98bdc3b52414967c964ead063ab1784168d2b76e/module-system/sov-modules-api/src/lib.rs#L424-L430

While the `Context` will always be the same for all dependencies in a module, their `GasConfig` will vary. It means this code would break when we introduce an associated type that is not the same for the whole set.

This is a restriction of the language. It will not allow us to accept `Any` generic argument there, even if we don't interact at all with the `GasConfig` on this implementation. It would be a problem to set the size of the type on other contexts, but not for us here.

Reply:
@bkolad we currently don't have a mechanism of implementing `sov-modules-api::Module` automatically via macro, as the `call` will be always customized for the module.

From my understanding, we should put all "metadata" into `ModuleInfo`, and have it implemented automatically (i.e. derive macro). The implication of moving `GasConfig` to `Module` is the user will have to define manually his `GasConfig`.

Reply:
I second this about having `GasConfig` being its own trait, similar to `CallMessage`

Reply:
As discussed on call, a couple of changes:

- move the `GasConfig` to `Module`
- move the `const` to `Default` implementation of `Module`
- create a derive macro helper attribute `gas` that will bind the type to the constant

Reply:
This is how the bank module will look like. Let me know your thoughts:

https://gist.github.com/vlopes11/c6413473b6c38d456ac07beba7512775